OsmAnd/src/net/osmand/plus/views/mapwidgets/widgets/AlarmWidget.java

```diff
@@ -138,29 +143,60 @@ public boolean updateInfo(DrawSettings drawSettings, boolean drawBitmap) {
 							icon.setImageResource(info.locImgId);
 						}
 					}
-					if (!Algorithms.objectEquals(info.text, cachedText) || cachedRegion != info.region) {
+
+					boolean isRegionChanged = cachedRegion != info.region;
+					if (isRegionChanged) {
+						cachedRegion = info.region;
+					}
+
+					boolean isAlarmTypeChanged = cachedAlarmType != alarm.getType();
+					if (isAlarmTypeChanged) {
+						cachedAlarmType = alarm.getType();
+					}
+
+					if (!Algorithms.objectEquals(info.text, cachedText) || isRegionChanged || isAlarmTypeChanged) {
 						changed = true;
 						cachedText = info.text;
-						cachedRegion = info.region;
 						if (layout != null && widgetText != null) {
 							widgetText.setText(cachedText);
 							Resources res = layout.getContext().getResources();
-							if (alarm.getType() == AlarmInfo.AlarmInfoType.SPEED_LIMIT && info.americanType && !info.isCanadianRegion) {
+							if (cachedAlarmType == AlarmInfo.AlarmInfoType.SPEED_LIMIT && info.americanType && !info.isCanadianRegion) {
 								int topPadding = res.getDimensionPixelSize(R.dimen.map_alarm_text_top_padding);
 								widgetText.setPadding(0, topPadding, 0, 0);
 							} else {
 								widgetText.setPadding(0, 0, 0, 0);
 							}
+
+							if (cachedAlarmType == AlarmInfo.AlarmInfoType.SPEED_CAMERA) {
+								FrameLayout.LayoutParams widgetTextLayoutParams = new FrameLayout.LayoutParams(
+										res.getDimensionPixelSize(R.dimen.map_alarm_speed_camera_speed_info_size),
+										res.getDimensionPixelSize(R.dimen.map_alarm_speed_camera_speed_info_size));
+
+								widgetTextLayoutParams.gravity = Gravity.RIGHT | Gravity.TOP;
+
+								widgetText.setLayoutParams(widgetTextLayoutParams);
+								widgetText.setGravity(Gravity.CENTER);
+								widgetText.setTextSize(COMPLEX_UNIT_PX, res.getDimensionPixelSize(R.dimen.map_alarm_speed_camera_speed_info_text_size));
+							} else {
+								FrameLayout.LayoutParams widgetTextLayoutParams = new FrameLayout.LayoutParams(
+										FrameLayout.LayoutParams.WRAP_CONTENT,
+										FrameLayout.LayoutParams.WRAP_CONTENT);
+
+								widgetTextLayoutParams.gravity = Gravity.CENTER;
+
+								widgetText.setLayoutParams(widgetTextLayoutParams);
+								widgetText.setGravity(Gravity.NO_GRAVITY);
+								widgetText.setTextSize(COMPLEX_UNIT_PX, res.getDimensionPixelSize(R.dimen.map_alarm_text_size));
+							}
 						}
 					}
-					if (!Algorithms.objectEquals(info.bottomText, cachedBottomText) || cachedRegion != info.region) {
+					if (!Algorithms.objectEquals(info.bottomText, cachedBottomText) || isRegionChanged || isAlarmTypeChanged) {
```

Comment:
If we talk about this code, then it's simple 2 blocks should be combined into 1 if, otherwise the pattern is broken and it's hard to maintain.

Reply:
Absolutely agree that single `if` will be simpler for understanding, will combine it. My initial intention was to reduce number of changes in PR.